### DISCUSSION TITLE
How to Implement Offline Timer Control for a Cyclic Switch

### DISCUSSION THREAD

**Discussion by Benlinmaker** [HomeSpan/HomeSpan#850] [Q&A] [answered]
Hi, First a big thank you to the Homespan team for their generous contributions

I am currently designing a multi-channel HomeKit light but have encountered an issue with offline control. I want to establish an offline button functionality, meaning that users can use a physical button to control the light with a timer function even without a Wi-Fi connection.

For example:
- **Single press:** The light turns on at 50% output, automatically turns off after 12 hours, and then automatically turns on again after another 12 hours.
- **Double press:** The light turns on at 100% output, automatically turns off after 12 hours, and then automatically turns on again after another 12 hours.
- If the power is cut off, no timer settings are remembered.

Currently, during testing, the physical button can turn the light on and off, but it does not turn off or on based on the timer. For quick testing, the cycle time is temporarily set to 3 seconds.

Where is the error in my timer settings? How should I modify them so that the LED lights can turn on and off endlessly?

Current development environment:

- M5stamp pico ESP32-d4
- self design PCB with 3 color LED
-  5v Fan
- D18B20 sensor

Here is my code

**Main sketch.ino**
```

#include "HomeSpan.h" 
#include "DEV_LED_W.h"
#include "DEV_LED_R.h"
#include "DEV_LED_G.h"
#include "DEV_LED_F.h"
#include "DEV_Sensors_18B20.h"       

void setup() {
  
  Serial.begin(115200);
    homeSpan.setControlPin(39); // sets control button
    homeSpan.setStatusPin(27); // sets LED Pin
    homeSpan.setApSSID("Luna-Setup");
    homeSpan.setApPassword("12345678");
    homeSpan.setPairingCode("33014431");
    homeSpan.enableAutoStartAP();
    homeSpan.begin(Category::Bridges,"Luna");
  
  new SpanAccessory();                            // This first Accessory is the new "Bridge" Accessory.  It contains no functional Services, just the Accessory Information Service
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
  
  new SpanAccessory(); 
  
    new Service::AccessoryInformation(); 
      new Characteristic::Identify();                
      new Characteristic::Name("白光光譜");//White
    new DEV_DimmableLED_W(32,39);                    //adjust pin 18 to pin 26, empty the pin 18 for Tempture sensor DS18B20

  new SpanAccessory(); 
  
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
      new Characteristic::Name("生長光譜"); //Red
    new DEV_DimmableLED_R(21,39);      

  new SpanAccessory(); 
  
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
      new Characteristic::Name("視覺光譜"); //Green
   new DEV_DimmableLED_G(33,39); 

  new SpanAccessory(); 
  
    new Service::AccessoryInformation();    
      new Characteristic::Identify();                       
      new Characteristic::Name("循環風扇"); //Fan
    new DEV_Fan(25,39); 

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("燈具溫度"); //DS18B20
    new DEV_TempSensor(22);
} // end of setup()

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();

} // end of loop()
```

**DEV_LED_W.h**
```

#include "extras/PwmPin.h" // allows PWM control of LED brightness

struct DEV_DimmableLED_W : Service::LightBulb { // Dimmable LED

  LedPin *ledPin; // Pin configuration details
  int powerPin; // NEW! pin with pushbutton to turn on/off LED
  SpanCharacteristic *power; // reference to the On Characteristic
  SpanCharacteristic *level; // reference to the Brightness Characteristic
  unsigned long lastButtonPressTime = 0; // Store the last button press time
  unsigned long lastLEDChangeTime = 0; // Store the last LED state change time
  bool isLEDOn = true; // LED state tracking

  DEV_DimmableLED_W(int pin, int powerPin) : Service::LightBulb() {
    power = new Characteristic::On();
    level = new Characteristic::Brightness(50); // Initial brightness
    level->setRange(5, 100, 1);
    new SpanButton(powerPin);
    this->powerPin = powerPin;
    this->ledPin = new LedPin(pin);
  }

  boolean update() {
    if (power->updated()) {
      // Time-based LED control
      unsigned long currentTime = millis(); // Assuming `timeVal()` returns system time in milliseconds
      if (currentTime - lastButtonPressTime >= 3000 && isLEDOn) { // 12 hours have passed 28800000
        power->setVal(false);
        lastLEDChangeTime = currentTime;
        isLEDOn = false;
      }
      if (!isLEDOn && currentTime - lastLEDChangeTime >= 6000) { // 12 hours have passed 57600000
        power->setVal(true);
        lastLEDChangeTime = currentTime;
        isLEDOn = true;
      }
      ledPin->set(power->getVal() * level->getVal());
    }
    return true;
  }

  void button(int pin, int pressType) override {
    if (pin == powerPin) {
      if (pressType == SpanButton::SINGLE || pressType == SpanButton::DOUBLE) {
        lastButtonPressTime = millis(); // Update the last button press time
        power->setVal(1 - power->getVal()); // Toggle the LED state
        isLEDOn = power->getVal();
        if (pressType == SpanButton::DOUBLE) {
          level->setVal(100); // Max brightness for double press
        } else {
          level->setVal(50); // Standard brightness for single press
        }
        ledPin->set(power->getVal() * level->getVal());
      }
    }
  }
};
```

**Comment by HomeSpan** [comment, marked as answer]
You've placed your timer logic in the `update()` portion of the code.  This is incorrect - it should be in the `loop()` portion of a Service.  See Tutorial Example 14 - Emulated Pushbuttons for a complete example.

**Reply by Benlinmaker** [reply]
I am extremely grateful for your guidance. Thank you for pointing out the programming error in this area. After modifying it according to your method, it started working correctly immediately! Thanks to the assistance of the Homespan team, I am deeply appreciative of your selfless dedication.

The following is the modified code that works correctly, provided for those who need a similar example.
'''
 bool buttonPressed = false;
  
void button(int pin, int pressType) override {
if(pin==powerPin){
    buttonPressed = true;
ledPin->set(power->getVal()*level->getVal());
      }
      };
      
void loop(){
    if (!buttonPressed) return;        // Only run the loop if the button has been pressed

    static unsigned long lastToggleTime = 0;
    static int lastLevel = 50; // Default level

    if (millis() - lastToggleTime >= 4000) { 
      lastToggleTime = millis();
      if (power->getVal()) {
        lastLevel = level->getVal(); // Save current level before turning off
        power->setVal(false);
      } else {
        power->setVal(true);
        level->setVal(lastLevel); // Restore last level
      }
      ledPin->set(power->getVal() * level->getVal());
    }
  }//loop
'
